File: prog2/Special/If.cs
// If -- Parse tree node strategy for printing the special form if

using System;

namespace Tree
{
    public class If : Special
    {
	public If() { }

        public override void print(Node t, int n, bool p)
        {
            Printer.printIf(t, n, p);
        }

        public override Node eval(Node t, Environment env) 
        {
          Node exp;
          Node cond = t.getCdr().getCar();
          if (cond.eval(env).isBool()) 
          {
            exp = t.getCdr().getCdr().getCar();
            return exp.eval(env);
          }
          else if (!t.getCdr().getCdr().getCdr().isNull()) 
          {
            exp = t.getCdr().getCdr().getCdr().getCar();
            return exp.eval(env);
          }
          Console.Error.WriteLine("Error in IF");
          return Nil.getInstance();
        }
    }
}

File: prog1/If.java
class If extends Special {
  void print(Node t, int n, boolean p) {
    String indent = " ".repeat(n);
    System.out.print(indent);

    if(!p) {
      System.out.print("(");
    }
    t.getCar().print(0);
    if(!t.getCdar().isNull()) {
      t.getCdar().print(1);
    }
    
    if(!t.getCddr().isNull()) {
      System.out.println();
      t.getCddr().print(n + 4, true);
    }
  }
}
